crates/core/src/network/partition_processor_rpc_client.rs

```diff
@@ -0,0 +1,270 @@
+// Copyright (c) 2024 - Restate Software, Inc., Restate GmbH.
+// All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use assert2::let_assert;
+use restate_types::identifiers::{PartitionProcessorRpcRequestId, WithPartitionKey};
+use restate_types::invocation::{InvocationQuery, InvocationResponse, ServiceInvocation};
+use restate_types::live::Live;
+use restate_types::net::partition_processor::{
+    GetInvocationOutputResponseMode, InvocationOutput, PartitionProcessorRpcRequest,
+    PartitionProcessorRpcRequestInner, PartitionProcessorRpcResponse, SubmitInvocationReplyOn,
+    SubmittedInvocationNotification,
+};
+use restate_types::partition_table::{FindPartition, PartitionTable, PartitionTableError};
+
+use crate::network::rpc_router::{RpcError, RpcRouter};
+use crate::network::NetworkSender;
+use crate::{my_node_id, ShutdownError};
+
+#[derive(Debug, thiserror::Error)]
+pub enum PartitionProcessorRpcClientError {
+    #[error(transparent)]
+    UnknownPartition(#[from] PartitionTableError),
+    #[error("failed sending request")]
+    SendFailed,
+    #[error(transparent)]
+    Shutdown(#[from] ShutdownError),
+    #[error("operation failed: {0}")]
+    Internal(String),
+}
+
+impl<T> From<RpcError<T>> for PartitionProcessorRpcClientError {
+    fn from(value: RpcError<T>) -> Self {
+        match value {
+            RpcError::SendError(_) => PartitionProcessorRpcClientError::SendFailed,
```

Comment:
I prefer to avoid using this pattern here, because i would have to leak abstractions back and forth and I don't really like it. I'm ok with cloning, eventually ServiceInvocation IMO should all go behind an arc or something (and already good part of it it's behind an arc). If this becomes a performance issue, i'll deal with it.